core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java

```diff
@@ -60,49 +108,68 @@ public SynonymTokenFilterFactory(IndexSettings indexSettings, Environment env, A
-            AnalysisRegistry.getSettingsFromIndexSettings(indexSettings, AnalysisRegistry.INDEX_ANALYSIS_TOKENIZER + "." + tokenizerName));
-        Analyzer analyzer = new Analyzer() {
-            @Override
-            protected TokenStreamComponents createComponents(String fieldName) {
-                Tokenizer tokenizer = tokenizerFactory == null ? new WhitespaceTokenizer() : tokenizerFactory.create();
-                TokenStream stream = ignoreCase ? new LowerCaseFilter(tokenizer) : tokenizer;
-                return new TokenStreamComponents(tokenizer, stream);
-            }
-        };
+    // for backward compatibility
+    /**
+     * @deprecated This filter tokenize synonyms with whatever tokenizer and token filters appear before it in the chain in 6.0.
+     */
+    @Deprecated
+    protected final TokenizerFactory tokenizerFactory;
+
+    public class Factory implements TokenFilterFactory{
+
+        private final String name;
+        private final SynonymMap synonymMap;
+
+        public Factory(String name, Analyzer analyzerForParseSynonym, Reader rulesReader) {
 
-        try {
-            SynonymMap.Builder parser = null;
+            this.name = name;
 
-            if ("wordnet".equalsIgnoreCase(settings.get("format"))) {
-                parser = new WordnetSynonymParser(true, expand, analyzer);
-                ((WordnetSynonymParser) parser).parse(rulesReader);
+            Analyzer analyzer;
+            if (tokenizerFactory != null) {
+                analyzer = new Analyzer() {
+                    @Override
+                    protected TokenStreamComponents createComponents(String fieldName) {
+                        Tokenizer tokenizer = tokenizerFactory.create();
+                        TokenStream stream = ignoreCase ? new LowerCaseFilter(tokenizer) : tokenizer;
+                        return new TokenStreamComponents(tokenizer, stream);
+                    }
+                };
             } else {
-                parser = new SolrSynonymParser(true, expand, analyzer);
-                ((SolrSynonymParser) parser).parse(rulesReader);
+                analyzer = analyzerForParseSynonym;
             }
 
-            synonymMap = parser.build();
-        } catch (Exception e) {
-            throw new IllegalArgumentException("failed to build synonyms", e);
+            try {
+                SynonymMap.Builder parser;
+                if ("wordnet".equalsIgnoreCase(format)) {
+                    parser = new WordnetSynonymParser(true, expand, analyzer);
+                    ((WordnetSynonymParser) parser).parse(rulesReader);
+                } else {
+                    parser = new SolrSynonymParser(true, expand, analyzer);
+                    ((SolrSynonymParser) parser).parse(rulesReader);
+                }
+                synonymMap = parser.build();
+            } catch (Exception e) {
+                throw new IllegalArgumentException("failed to build synonyms", e);
+            }
```

Comment:
I think we should close the analyzer if tokenizerFactory is not null. Otherwise, it is never closed.

```
finally {
    if (tokenizerFactory != null) {
        analyzer.close();
    }
}
```

(this looks like a pre-existing bug, it was not introduced by your PR)